### DISCUSSION TITLE
Hey! Ideas here.

### DISCUSSION THREAD

**Discussion by TwoAbove** [TwoAbove/noita-tools#73] [Ideas]
Hello! 

If you have ideas on how to improve this tool, add them here.

Anything from UI to QoL to new features.

**Comment by TwoAbove** [comment]
https://github.com/TwoAbove/noita-tools/issues/66

Nightmare mode seed info

**Reply by TwoAbove** [reply]
This will just need an update to the x,y coords of holy mountains in nightmare

**Comment by TwoAbove** [comment]
https://github.com/TwoAbove/noita-tools/issues/65

Show wands and spells in holy mountains

**Reply by TwoAbove** [reply]
This requires decompiling and reverse-engineering the in-engine `GetRandomActionWithType` and `GetRandomAction` functions. I did not get very far, since I don't have that much experience with reverse-engineering.  
Maybe a more semantic approach through gathering statistics through a lua mod might be a good idea? Then I can get some sense of what goes on inside.
Another idea is to hook the exe and run a lua mod that runs the gen command once, and I can run debugging on it.
Another idea is to reach out to the devs and ask for the `falling_everything.pdb` that will add lots of mapping to the debugging info.
Even more bold, I can just ask for a description of how these functions work from the devs :smile:

**Reply by TwoAbove** [reply]
I've almost cracked the way these work!

**Reply by TwoAbove** [reply]
This needs a bit more work. Looks like lua -> ts discrepancy

**Comment by TwoAbove** [comment]
https://github.com/TwoAbove/noita-tools/issues/59

Calculate gamble perk shift

**Reply by TwoAbove** [reply]
I'll start doing this one, seems like a quick win

**Comment by Breezebuilder** [comment]
This is the best seed summary tool, thanks for your hard work.

A dark mode toggle would be great. Switching between the dark caves of Noita and a white browser screen can be a bit of a shock to the eyes.

**Reply by TwoAbove** [reply]
This is a good and not very had one to implement!

I'll add it to the feature update (that adds the gamble shift)

**Reply by Breezebuilder** [reply]
Thanks!

Name tooltips for the perk images would be a handy QoL feature too, I haven't played for long enough to memorise all the icons.
Looks like the names are already in the `img alt` attribute, just need to include it as a `title` attribute too.

**Reply by TwoAbove** [reply]
@breezebuilder 

Dark theme is in the dev deployment. Will be live this evening

**Reply by TwoAbove** [reply]
You can check it out https://noita-tools-dev.herokuapp.com/

**Reply by Breezebuilder** [reply]
Great work, thanks!

**Comment by Roxolasu** [comment]
Another nice feature would be a perk gamble preview.

Logic for perk gamble (perk.lua, at the end of perk_pickup):
```lua
SetRandomSeed(perkX, perkY)
if(Random(1,100)<=PERK_DESTROY_CHANCE) then
    destroyPerk()
end
```

Perk positions are fixed: (perk.lua, end of perk_spawn_many)
- Y is fixed per Temple (1410, 2946, 6530, 8578, ...)
- X is based on the number of perks:
  - Confirmed in-game:
    - 3 Perks: -22, -2, 18
    - 4 Perks: -24, -10, 6, 20
    - 5 Perks: -26, -14, -2, 10, 22
  - FYI: There is another unrelated entity at X = -2, don't get confused about a different Y-Height
  - In general: -32 + (PERK_INDEX_START_1 - 0.5) * (60 / PERK_COUNT)
    - Perk positions seem to be whole numbers, I'm not sure where / how they are rounded/cutoff
    - Rounding is strange for 4 perks: Calculated locations are -24.5, -9.5, 5.5, 20.5

For example, to check if the perk in the leftmost slot in the first temple will make the others disappear (assuming one has 1x perk lottery):
```lua
SetRandomSeed(-22, 1410)
if(Random(1,100)<=50) then
    destroyPerk()
end
```

**Reply by TwoAbove** [reply]
This is awesome! 

I'll definitely add this, but will need to think about how to implement it in the UI

**Reply by TwoAbove** [reply]
@Roxolasu deployed!

**Reply by Roxolasu** [reply]
Great, thank you!

**Reply by Roxolasu** [reply]
@TwoAbove: There seems to be a slight implementation bug, your perkNumber starts at 0, whereas Noita starts counting at 1. Therefore, your calculated coordinates are -42,-22,-2 instead of -22, -2, 18.

**Reply by TwoAbove** [reply]
I'll make a fix right away

**Reply by TwoAbove** [reply]
@Roxolasu I wouldn't call it slight :smile: It's now fixed

**Reply by Roxolasu** [reply]
That was fast. Unfortunately there is still a problem for 4 or 7+ perks: 
The calculated positions are rounded to whole integers when the perk entity is spawned. This rounded position is used for the lottery. Therefore, our perk numbers need to be rounded too. I cant seem to find a function which matches all observed values, its not a simple cast, nor Math.floor(x+0.5), neither Math.ceil(x-0.5).

My best approach would be to hardcode the two problematic values for 4 perks, and ignore potential problems at 11+perks:
``` javascript
function castToInt(x){
	switch(x){
            // special cases for exactly 4 perks.            
            case -9.5: return -10;
            case 20.5: return 20;
            default: return Math.floor(x+0.5)           
  }
}
// testvector for 4, 7 and 8 perks, generated by comparing calculated positions with positions observed in-game
console.log(castToInt(-24.5) == -24,castToInt(-9.5) == -10,castToInt(5.5) == 6,castToInt(20.5) == 20 ,castToInt(-27.71) == -28,castToInt(-19.14) == -19,castToInt(-10.57) == -11,castToInt(6.57) == 7,castToInt(15.14) == 15,castToInt(23.71) == 24,castToInt(-28.25) == -28,castToInt(-20.75) == -21,castToInt(-13.25) == -13,castToInt(-5.75) == -6,castToInt(1.75) == 2,castToInt(9.25) == 9,castToInt(16.75) == 17,castToInt(24.25) == 24)
```

**Reply by TwoAbove** [reply]
Yeah, looks like that's one way of doing it. I was also thinking about float -> int downcasting working, but that might be difficult to do with js, and I'm not yet sure if that might be the solution. 

In the mean time, I'll just hard-code it

**Reply by TwoAbove** [reply]
@Roxolasu  Updated

**Reply by Roxolasu** [reply]
Another minor fix, and it should be working :). See #82

**Reply by TwoAbove** [reply]
Hey @Roxolasu How did you get the pixel rounding for perks? I'm having a simiar issue with wands and spells – looks like there are some funky float/int transitions going on

**Reply by Roxolasu** [reply]
@TwoAbove I used the debug mode (from the modding tools) to check the actual position of the entities.
(Start Noita-dev, f5 to enable debug, Press I + click on the relevant entity to see its position). [A bit more info](https://steamcommunity.com/app/881100/discussions/0/3476150824940419212/)

The corresponding rounding mode seems to be [Round half to even](https://en.wikipedia.org/wiki/Rounding#Round_half_to_even ). If you implement it, you could use the above console.log statements to test it

**Reply by TwoAbove** [reply]
Hm, that's an interesting idea! Thanks!

**Comment by TrabacchinLuigi** [comment]
Allow to select more than one perk from the same holy mountain when they are marked from perk lottery

**Reply by TwoAbove** [reply]
Thanks for the idea! 

I'm already working on that. It'll take a bit of time to redo how things are handled on the inside

**Reply by TwoAbove** [reply]
@TrabacchinLuigi It should be deployed shortly

**Comment by TrabacchinLuigi** [comment]
Recently used seeds, maybe in local storage, and maybe the list should save and display starting items too.
Or maybe a preferred list, with a button to pin or star a seed Wich let you add a comment to it

**Reply by TwoAbove** [reply]
One thing that I'm worried about is that storing info like that adds expectations about its persistence, and safari, for example, deletes local storage periodically. 

One solution is to add accounts and store this in a db, but that will require a db to be available. The current server is stateless and does not have a db, and ideally, I'd like to keep it that way.

If you have any ideas as to how to do this, I'll happily take them into account

**Reply by TrabacchinLuigi** [reply]
Periodically may be more than the user expectation, a small disclaimer could be enough to tone down the expectations, i would not invest in such an implementation unless you go for making lot of users, but this does not look like a project that could interest investors...

**Reply by TwoAbove** [reply]
I'm not planning for investors ever, and this is more of a passion project that I'm more than willing to spend another $5-10/month on a database. 

A disclaimer feels like it would be a solution, but I don't know if that's something I'd be satisfied with.

**Reply by TrabacchinLuigi** [reply]
Then i think you have an answer on the way to go :)
I would suggest authentication only to use the feature then, maybe from some sso

**Comment by JooJerOne** [comment]
I dont know if the code to calculate how Alchemic Precursor and Lively Concoction has already been found for this version, but if it did, add it to the site to make it more complete

**Reply by TwoAbove** [reply]
Hey! There is a toggle to enable alchemy in the footer

**Reply by JooJerOne** [reply]
Oh my god, im sorry, i was a dummy and didnt see the "enable alchemy spoilers" button

Thanks for your work, Seva!

**Comment by TrabacchinLuigi** [comment]
i'm not really sure about this one:

Re-orderable holy mountains:

Many players go visit other holy mountains before completing the main one, so the order of the perks gets confused (this should be verified) ...
Would be nice to have some way to specify that, maybe as a list of squares with the notation "MWn" "E1n" "W2n" where n is the depth of the mountain

**Reply by TwoAbove** [reply]
I think this would just add confusion. Since app state is not durable, even browser tab freezing might clear it, making it unreliable at best, and frustrating at least. 

It really boils down to why this feature might exist, and what expectations users would have about the feature. 

Here, I think that this feature will need a lot of extra code that will need support to make it meet minimal expectations, and also might just add confusion to the usage of everything else

**Comment by TrabacchinLuigi** [comment]
another feature that would be very nice is the ability to search decks,
Also this one is to be verified, but i belive some perks and spell get removed in the seed.
Instead of searching the specific level or the starting spell, search if a perk is present or a spell is present

**Reply by TwoAbove** [reply]
This is a great one! I'll add it to the list of things I really want to add

**Comment by twothe** [comment]
Dark Mode would really be a great thing to add.

**Reply by TwoAbove** [reply]
Hey! You can enable dark mode by toggling the dark mode toggle in the footer of the page

**Reply by twothe** [reply]
Thank you, found it. :D

**Comment by H4x0t** [comment]
Big ask, but just an idea.... Stream Deck implementation of your code. It would be very cool to see the need to know info at a glance when I restart a seed.

**Reply by TwoAbove** [reply]
Hey! What would that be like? I don't have a steam deck

**Reply by TwoAbove** [reply]
If it has a browser, then it might work

**Comment by TwoAbove** [comment]
I'm locking this thread because it's getting confusing. Please add new issues here: https://github.com/TwoAbove/noita-tools/issues/
